pkg/cmd/monitor/monitor.go

```diff
@@ -0,0 +1,279 @@
+	fs.StringVar(&o.clientKeyFile, "key-file", "", "Health probe TLS client key file. (required)")
+	fs.StringVar(&o.clientCACertFile, "cacert-file", "", "Health probe TLS client CA certificate file. (required)")
+}
+
+var shutdownSignals = []os.Signal{os.Interrupt, syscall.SIGTERM}
+
+func (o *monitorOpts) Validate(_ context.Context) error {
+	if len(o.clientKeyFile) == 0 {
+		return errors.New("missing required flag: --key-file")
+	}
+	if len(o.clientCertFile) == 0 {
+		return errors.New("missing required flag: --cert-file")
+	}
+	if len(o.clientCACertFile) == 0 {
+		return errors.New("missing required flag: --cacert-file")
+	}
+	return nil
+}
+
+func (o *monitorOpts) Run(ctx context.Context) error {
+	defer utilruntime.HandleCrash()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// handle teardown
+	shutdownHandler := make(chan os.Signal, 2)
+	signal.Notify(shutdownHandler, shutdownSignals...)
+	go func() {
+		select {
+		case <-shutdownHandler:
+			klog.Infof("Received SIGTERM or SIGINT signal, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		case <-ctx.Done():
+			klog.Infof("Context has been cancelled, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		}
+	}()
+
+	// enable profiler by default
+	defer serviceability.Profile("web").Stop()
+	serviceability.StartProfiler()
+
+	lg, err := health.GetZapLogger(health.LoglevelToZap(o.logLevel), o.logOutputs, o.enableLogRotation, o.LogRotationConfigJSON)
+	if err != nil {
+		return err
+	}
+
+	// TODO make health checks configurable by flag(s)
+	monitor, err := o.newMonitor(ctx, lg,
+		WithSingleTargetHealthCheck(
+			health.WithSerializedReadSingleTarget(),
+			health.WithGRPCReadySingleTarget(),
+			health.WithQuorumReadSingleTarget(),
+			// TODO
+			// health.WithWatchRequireLeader(),
+		),
+		WithMultiTargetHealthCheck(
+			health.WithQuorumRead(),
+			// TODO
+			// health.WithNodeICMP(),
+		),
+	)
+
+	monitor.wg.Add(1)
+	go func() {
+		defer monitor.wg.Done()
+		monitor.Schedule()
+	}()
+	monitor.wg.Wait()
+	lg.Info("health monitor is shutting down",
+		zap.String("pod", o.podName),
+		zap.Int("static-pod-version", o.staticPodVersion),
+	)
+	return nil
+}
+
+// newMonitor creates a series of health checks. Health check probes can be single or multi target.
+func (o *monitorOpts) newMonitor(ctx context.Context, lg *zap.Logger, singleTargetChecks []health.CheckFunc, multiTargetChecks []health.CheckFunc) (*Monitor, error) {
+	var healthChecks []*health.Check
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      o.clientCertFile,
+		KeyFile:       o.clientKeyFile,
+		TrustedCAFile: o.clientCACertFile,
+	}
+
+	targets := strings.Split(o.Targets, ",")
+
+	// Create single target checks one check per target
+	for _, target := range targets {
+		// one client per target to eliminate lock racing
+		client, err := newETCD3Client(tlsInfo, targets)
+		if err != nil {
+			return nil, err
+		}
+		for _, healthCheckFunc := range singleTargetChecks {
```

Comment:
I expected to see the client created here to avoid re-use.

Reply:
the idea is 1 client per target and then each target has multiple health checks using same target. I wanted to reduce the number of clients but I will check perf both ways.